### PR TITLE
sensor: tdk 42688 driver

### PR DESCRIPTION
**Pull request by yperess**
Adds a driver for TDK InvenSense 42688 six axis IMU. Verified using the sensor shell sample app via:

sensor info
sensor get icm42688p@0
Collaborated on by Tom Burdick and Yuval Peress

### CODE REVIEW

**Review by teburd — DISMISSED** [review]
I'm good with this, I have some patches I can PR once in.

**Review by teburd — DISMISSED** [review]

File: dts/bindings/sensor/invensense,icm42688.yaml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+description: ICM-42688 motion tracking device
+
+compatible: "invensense,icm42688"
+
+include: [sensor-device.yaml, spi-device.yaml]
+
+properties:
+    friendly-name:
+      default: "ICM42688"
```
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42688.yaml:14]: I don't think it makes sense to set a default friendly name in the binding; especially if that default is just the model name.

File: dts/bindings/sensor/invensense,icm42688.yaml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        property value should ensure the flags properly describe the
+        signal that is presented to the driver.
+
+    accel-hz:
+      type: int
+      required: true
+      description: |
+        Default frequency of accelerometer. (Unit - Hz)
+        Maps to ACCEL_ODR field in ACCEL_CONFIG0 setting
+        Power-on reset value is 1000.
```
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42688.yaml:28]: Make the property optional and set the default to the power-on reset value

File: dts/bindings/sensor/invensense,icm42688.yaml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        - 16000
+        - 32000
+
+    gyro-hz:
+      type: int
+      required: true
+      description: |
+        Default frequency of gyroscope. (Unit - Hz)
+        Maps to GYRO_ODR field in GYRO_CONFIG0 setting
+        Power-on reset value is 1000.
```
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42688.yaml:52]: Make the property optional and set the default to the power-on reset value

File: dts/bindings/sensor/invensense,icm42688.yaml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        - 16000
+        - 32000
+
+    accel-fs:
+      type: int
+      required: true
+      description: |
+        Default full scale of accelerometer. (Unit - g)
+        Maps to ACCEL_FS_SEL field in ACCEL_CONFIG0 setting
+        Power-on reset value is 16
```
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42688.yaml:76]: Make the property optional and set the default to the power-on reset value

File: dts/bindings/sensor/invensense,icm42688.yaml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        - 4
+        - 2
+
+    gyro-fs:
+      type: int
+      required: true
+      description: |
+        Default full scale of gyroscope. (Unit - DPS)
+        Maps to GYRO_FS_SEL field in GYRO_CONFIG0 setting
+        Power-on reset value is 2000
```
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icm42688.yaml:89]: Make the property optional and set the default to the power-on reset value

File: samples/sensor/sensor_shell/prj.conf [hunk 6]
```diff
@@ -1,5 +1,7 @@
 CONFIG_SHELL=y
-CONFIG_BOOT_BANNER=n
+CONFIG_LOG=y
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply samples/sensor/sensor_shell/prj.conf:2]: Agree they should be dropped

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add the sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI

File: drivers/sensor/icm42688/icm42688.c [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+#include <zephyr/drivers/spi.h>
+#include <zephyr/sys/byteorder.h>
+#include "icm42688.h"
+#include "icm42688_reg.h"
+#include "icm42688_spi.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ICM42688_LL, CONFIG_SENSOR_LOG_LEVEL);
+
+void icm42688_lock(const struct device *dev)
```
Comment (keith-zephyr) [inline drivers/sensor/icm42688/icm42688.c:19]: Why do you have an empty lock function?

File: drivers/sensor/icm42688/icm42688_core.c [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+static const struct sensor_driver_api icm42688_driver_api = {
+	.sample_fetch = icm42688_sample_fetch,
+	.channel_get = icm42688_channel_get,
+	.attr_set = icm42688_attr_set,
+	.attr_get = icm42688_attr_get,
+};
+
+int icm42688_init(const struct device *dev)
```
Comment (keith-zephyr) [inline drivers/sensor/icm42688/icm42688_core.c:152]: Nit - I'd prefer that the driver implementation is provided by the module matching the sensor name.  So this file would be `icm42688.c` and the file with the helper routines `icm42688_reset()` and `icm42688_configure()` would be called `icm42688_core.c` or maybe `icm42688_support.c`.

@MaureenHelm what your thoughts?

File: boards/arm/tdk_robokit1/tdk_robokit1_defconfig [hunk 9]
```diff
@@ -12,6 +12,8 @@ CONFIG_SERIAL=y
 CONFIG_BOARD_TDK_ROBOKIT1=y
 CONFIG_WDT_DISABLE_AT_BOOT=y
 CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_SPI=y
+CONFIG_SPI_SAM=y
```
[the remark this answers is not in the document]
Comment (yperess) [reply boards/arm/tdk_robokit1/tdk_robokit1_defconfig:16]: Done

File: drivers/sensor/icm42688/Kconfig [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright (c) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_ICM42688 := invensense,icm42688
+
+config ICM42688
+	bool "ICM42688 Six-Axis Motion Tracking Device"
+	default $(dt_compat_enabled,$(DT_COMPAT_ICM42688))
+	depends on SPI
```
[the remark this answers is not in the document]
Comment (yperess) [reply drivers/sensor/icm42688/Kconfig:12]: Thanks, that's cool, didn't know about the `DT_HAS_..._ENABLED` macro

File: drivers/sensor/icm42688/icm42688.h [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+ * @brief The value of gravitational constant in micro m/s^2.
+ * TODO this is redefined because sensor v2 redefines SENSOR_G
+ */
+#define ICM42688_G 9806650LL
+
+/**
+ * @brief The value of constant PI in micros.
+ * TODO this is redefined because sensor v2 redefines SENSOR_PI
+ */
+#define ICM42688_PI 3141592LL
```
[the remark this answers is not in the document]
Comment (yperess) [reply drivers/sensor/icm42688/icm42688.h:285]: Done

File: dts/bindings/sensor/invensense,icm42688.yaml (hunk 1 above)
Comment (yperess) [reply dts/bindings/sensor/invensense,icm42688.yaml:14]: Done

File: dts/bindings/sensor/invensense,icm42688.yaml (hunk 3 above)
Comment (yperess) [reply dts/bindings/sensor/invensense,icm42688.yaml:52]: Done

File: dts/bindings/sensor/invensense,icm42688.yaml (hunk 5 above)
Comment (yperess) [reply dts/bindings/sensor/invensense,icm42688.yaml:89]: Done

File: dts/bindings/sensor/invensense,icm42688.yaml (hunk 4 above)
Comment (yperess) [reply dts/bindings/sensor/invensense,icm42688.yaml:76]: Done

File: samples/sensor/sensor_shell/prj.conf (hunk 6 above)
[the remark this answers is not in the document]
Comment (yperess) [reply samples/sensor/sensor_shell/prj.conf:2]: Done

File: drivers/sensor/icm42688/icm42688.c (hunk 7 above)
Comment (yperess) [reply drivers/sensor/icm42688/icm42688.c:19]: It was used in Tom's additional changes. I'll remove for now until it's needed.

File: drivers/sensor/icm42688/icm42688_core.c (hunk 8 above)
Comment (yperess) [reply drivers/sensor/icm42688/icm42688_core.c:152]: I ended with `icm42688_common.c` and `icm42688.c`, does that work?

**Comment by yperess** [comment]
> Please add the sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI

Done

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 12]
```diff
@@ -41,90 +41,90 @@ test_i2c_iaqcore: iaqcore@5 {
 	reg = <0x5>;
 };
 
-test_i2c_bme280: bme280@7 {
+test_i2c_bme280: bme280@6 {
```
Comment (keith-zephyr) [inline tests/drivers/build_all/sensor/i2c.dtsi:44]: Nit - add a comment to the file header indicating that the I2C peripheral addresses should be sequential.

File: tests/drivers/build_all/sensor/spi.dtsi [hunk 13]
```diff
@@ -26,9 +26,9 @@ test_spi_apa102: apa102@2 {
 	spi-max-frequency = <0>;
 };
 
-test_spi_rf2xx: rf2xx@4 {
+test_spi_rf2xx: rf2xx@3 {
```
Comment (keith-zephyr) [inline tests/drivers/build_all/sensor/spi.dtsi:29]: Same here - document that the SPI addresses should be sequential.

File: tests/drivers/build_all/sensor/spi.dtsi [hunk 14]
```diff
@@ -318,8 +318,20 @@ test_spi_icm42670: icm42670@3c {
...
 
-test_spi_bme680: bme680@3c {
+test_spi_bme680: bme680@27 {
 	compatible = "bosch,bme680";
-	reg = <0x3c>;
+	reg = <0x27>;
 	spi-max-frequency = <0>;
 };
+
+/* Requires the SPI to be atml,sam-spi
```
Comment (keith-zephyr) [inline tests/drivers/build_all/sensor/spi.dtsi:327]: I assume you meant to submit this as a separate commit?

**Comment by yperess** [comment]
> > Please add the sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI
> 
> @MaureenHelm I tried this but ran into a few issues (see the second commit in this PR)
> 
> 1. The test wasn't passing due to conflicting addresses (I fixed that)
> 2. The driver requires the SPI_SAM (@teburd can you confirm?) which can't be built for native_posix, so I can't seem to add it to this test without major refactoring.

Discussed with @teburd offline and there's no hard requirement for the SPI_SAM, CI should pass now

**Review by keith-zephyr — DISMISSED** [review]

File: tests/drivers/build_all/sensor/spi.dtsi (hunk 14 above)
Comment (yperess) [reply tests/drivers/build_all/sensor/spi.dtsi]: I left it in there so Tom & Maureen could see what I tried. It's been split up into a separate commit now that it works.

File: tests/drivers/build_all/sensor/spi.dtsi (hunk 13 above)
Comment (yperess) [reply tests/drivers/build_all/sensor/spi.dtsi:29]: Done

File: tests/drivers/build_all/sensor/i2c.dtsi (hunk 12 above)
Comment (yperess) [reply tests/drivers/build_all/sensor/i2c.dtsi:44]: Done

**Review by teburd — DISMISSED** [review]

**Review by keith-zephyr — DISMISSED** [review]

File: drivers/sensor/icm42688/icm42688.h [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+	uint32_t uc, uxa, uya, uza, uxg, uyg, uzg;
+
+	int16_t raw[7];
+
+	for (int i = 0; i < 7; i++) {
+		raw[i] = sys_le16_to_cpu((data[i * 2] << 8) | data[i * 2 + 1]);
+	}
+
+	printk("temp raw %d, accel x %d, y %d, z %d, gyro x %d, y %d, z %d\n", raw[0], raw[1],
+	       raw[2], raw[3], raw[4], raw[5], raw[6]);
```
Comment (MaureenHelm) [inline drivers/sensor/icm42688/icm42688.h:394]: `LOG_DBG`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Minor nit, but otherwise looks good. Also needs a rebase

File: drivers/sensor/icm42688/icm42688.h (hunk 15 above)
Comment (yperess) [reply drivers/sensor/icm42688/icm42688.h:394]: Done

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]
